addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/http/ComparableResponse.java

```diff
@@ -0,0 +1,529 @@
+     * heuristics that have low similarity between the original response and the reference response.
+     *
+     * @param referenceResponse the response to compare with, one that should have similar results.
+     */
+    public void tuneHeuristicsWithResponse(ComparableResponse referenceResponse) {
+        bodyStructureWeight =
+                bodyTreesStructureHeuristic(this, referenceResponse) * bodyStructureWeight;
+        headersWeight = headersCompareHeuristic(this, referenceResponse) * headersWeight;
+        numberLinesWeight = lineCountHeuristic(this, referenceResponse) * numberLinesWeight;
+        numberWordsWeight = wordCountHeuristic(this, referenceResponse) * numberWordsWeight;
+        reflectionWeight = inputReflectionHeuristic(this, referenceResponse) * reflectionWeight;
+        relevantKeywordsWeight =
+                relevantKeywordsCountHeuristic(this, referenceResponse) * relevantKeywordsWeight;
+    }
+
+    /**
+     * Compares two responses.
+     *
+     * @param response1 The response to compare with response2
+     * @param response2 The response compare with response1
+     */
+    public static float compareMessages(
+            ComparableResponse response1, ComparableResponse response2) {
+        return response1.compareWith(response2);
+    }
+
+    /**
+     * Return a heuristic evaluation between the status code of the 2 responses. 0 means very
+     * different and 1 very similar.
+     *
+     * @param response1 The first response to compare
+     * @param response2 The second response to compare
+     */
+    public static float statusCodeHeuristic(
+            ComparableResponse response1, ComparableResponse response2) {
+        return response1.statusCode != response2.statusCode ? 0f : 1f;
+    }
+
+    /**
+     * Return a heuristic evaluation between the number relevant keywords in the 2 responses. 0
+     * means very different and 1 very similar.
+     *
+     * @param response1 The first response to compare
+     * @param response2 The second response to compare
+     */
+    public static float relevantKeywordsCountHeuristic(
+            ComparableResponse response1, ComparableResponse response2) {
+        float nRelevantKeywordsR1 = 0;
+        float nRelevantKeywordsR2 = 0;
+        for (String keyword : RELEVANT_KEYWORDS) {
+            if (keyword.equalsIgnoreCase(response1.getValueSent())
+                    || keyword.equalsIgnoreCase(response2.getValueSent())) {
+                continue;
+            }
+            nRelevantKeywordsR1 +=
+                    StringUtils.countMatches(response1.body.toLowerCase(), keyword.toLowerCase());
+            nRelevantKeywordsR2 +=
+                    StringUtils.countMatches(response2.body.toLowerCase(), keyword.toLowerCase());
+        }
+
+        if (nRelevantKeywordsR1 > nRelevantKeywordsR2) {
+            return nRelevantKeywordsR2 / nRelevantKeywordsR1;
+        } else if (nRelevantKeywordsR1 < nRelevantKeywordsR2) {
```

Comment:
There are a lot of places that the else could be removed.

Reply:
If needs to check before returning 1f in the end? (Unless I'm totally missing something.)